File: Chapter10/VB/VBProject_10/AdvancedMedia/3-Webcam/WebCam/MemoryStreamAudioSink.vb
Imports System
Imports System.Net
Imports System.Windows
Imports System.Windows.Controls
Imports System.Windows.Documents
Imports System.Windows.Ink
Imports System.Windows.Input
Imports System.Windows.Media
Imports System.Windows.Media.Animation
Imports System.Windows.Shapes
Imports System.IO

' Namespace WebCam

Public Class MemoryStreamAudioSink
    Inherits AudioSink

    Protected Overrides Sub OnCaptureStarted()

        _stream = New MemoryStream()
    End Sub '   OnCaptureStarted

    Protected Overrides Sub OnCaptureStopped()

    End Sub '   OnCaptureStopped

    Public ReadOnly Property oAudioFormat() As AudioFormat
        Get
            Return (_audioFormat)
        End Get
    End Property

    Public ReadOnly Property AudioData() As MemoryStream
        Get

            Return (_stream)
        End Get
    End Property

    Protected Overrides Sub OnFormatChange(oAudioFormat As AudioFormat)

        If (Me._audioFormat Is Nothing) Then
            Me._audioFormat = oaudioFormat
        Else
            Throw New InvalidOperationException()
        End If
    End Sub '   OnFormatChange

    Protected Overrides Sub OnSamples(sampleTime As Long, sampleDuration As Long, sampleData As Byte())

        _stream.Write(sampleData, 0, sampleData.Length)
    End Sub '   OnSamples

    Dim _stream As MemoryStream
    Dim _audioFormat As AudioFormat
End Class   '   MemoryStreamAudioSink
' End Namespace   '   WebCam
' ..\Project_10\AdvancedMedia\3-Webcam\WebCam\MemoryStreamAudioSink.cs

File: Chapter10/C#/Project/AdvancedMedia/3-Webcam/WebCam/MemoryStreamAudioSink.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;
using System.IO;

namespace WebCam
{
    public class MemoryStreamAudioSink : AudioSink
    {
        protected override void OnCaptureStarted()
        {
            stream = new MemoryStream();
        }
        protected override void OnCaptureStopped()
        {
        }
        public AudioFormat AudioFormat
        {
            get
            {
                return (audioFormat);
            }
        }
        public MemoryStream AudioData
        {
            get
            {
                return (stream);
            }
        }
        protected override void OnFormatChange(AudioFormat audioFormat)
        {
            if (this.audioFormat == null)
            {
                this.audioFormat = audioFormat;
            }
            else
            {
                throw new InvalidOperationException();
            }
        }
        protected override void OnSamples(long sampleTime, long sampleDuration, byte[] sampleData)
        {
            stream.Write(sampleData, 0, sampleData.Length);
        }
        MemoryStream stream;
        AudioFormat audioFormat;
    } 
}
